app/src/main/java/com/github/polypoly/app/models/game/GameViewModel.kt

```diff
@@ -43,10 +63,160 @@ class GameViewModel(
         return gameEndedData
     }
 
+    fun getPlayerStateData(): LiveData<PlayerState> {
+        return playerStateData
+    }
+
+    private suspend fun gameLoop() {
+        var currentGame = gameData.value
+
+        while (currentGame != null && !currentGame.isGameFinished()) {
+            playerStateData.value = PlayerState.ROLLING_DICE
+            setLoading(false)
+
+            delay(currentGame.rules.roundDuration.toLong() * 1000 * 60)
+
+            playerStateData.value = PlayerState.TURN_FINISHED
+
+            nextTurn()
+
+            currentGame = gameData.value
+        }
+    }
+
+    /**
+     * Computes the next turn state and synchronizes with the other players
+     */
     fun nextTurn() {
-        gameData.value?.nextTurn()
-        roundTurnData.value = gameData.value?.currentRound ?: -1
-        gameEndedData.value = gameData.value?.isGameFinished() ?: false
+        setLoading(true)
+        viewModelScope.launch {
+            gameData.value?.nextTurn()
+
+            synchronizeGame().thenApply { syncSucceeded ->
+                if (syncSucceeded) {
+                    roundTurnData.value = gameData.value?.currentRound ?: -1
+                    gameEndedData.value = gameData.value?.isGameFinished() ?: false
+                }
+                setLoading(false)
+            }
+        }
+    }
+
+    private fun synchronizeGame(): CompletableFuture<Boolean> {
+        val gameUpdated = gameData.value ?: return CompletableFuture.completedFuture(false)
+        return remoteDB.getValue<Game>(gameUpdated.key).thenCompose { gameFound ->
+            if (gameFound.currentRound < gameUpdated.currentRound) {
+                gameData.value = gameUpdated
+                remoteDB.setValue(gameUpdated)
+            } else { // up to date version already on live db
+                gameData.value = gameFound
+                CompletableFuture.completedFuture(true)
+            }
+        }
+    }
+
+    private fun playerStateFSMTransition(expectedFrom: PlayerState, to: PlayerState) {
+        if (playerStateData.value != expectedFrom)
```

Comment:
The case where the current player state in playerStateData is different from expectedFrom is an error that should never have happened? Because if so, maybe throw an exception instead

Reply:
I initially didn't want to have such a hard end upon wrong transitions, but it should indeed invalidate the fsm a bit more than by just ignoring what happened